docs/source/miscellaneous/doc/adr/0016-use-tickets-encoding-for-transactions.md

```diff
@@ -499,7 +499,131 @@ the worker pool has a finite size and even with the full batching algorithm in t
+The values themselves are stored as dynamic columns with a var-long dynamic column key.
+
+- The value stored at column 0 is a ``SequenceAndBound`` which indicates the sequence number of the currently valid
+  value and the validity bound.
+- Values are stored at column IDs equal to a fresh timestamp at the time they're written.
+
+The main reason for this scheme is that while the validity bound itself changes frequently (approximately every
+``ADVANCEMENT_QUANTUM`` timestamps, which is 5,000,000 at time of writing), the value itself changes much less often.
+For example, with transactions2, in the absence of rollbacks, the value is expected to change just once in the
+service's lifetime, keeping track of the point where reading from transactions2 should begin.
+To further support this scheme, the coordination service recognises identity transformations from the current value,
+and doesn't write a new value in these cases.
+
+We have not implemented cleanup of the coordination service at this time, as we don't expect the number of values
+in a coordination sequence to be large. Furthermore, we don't ever range scan this table.
+
+#### Determining Transaction Schema Versions
+
+We still need to retain the ability to read from the old transactions table, when checking if/when values written
+before the transactions2 migration occurred. Thus, we replace the old ``TransactionService`` (that only reads/writes
+from transactions1) with one that delegates between multiple TransactionServices (concretely, just V1 and V2 at this
+time) depending on the specific timestamps involved in calls.
+
+The mechanism for determining which ``TransactionService`` to use goes through the aforementioned
+``CoordinationService``. We agree on a ``TimestampPartitioningMap``, which is a mapping of timestamps to transaction
+schema versions. This is a range map, as we want to support rollbacks easily. The map also is guaranteed to span the
+ranges [1, +∞) and be connected; note that this does not mean future behaviour is fixed, since in practice this map
+is read together with a validity bound. Thus, even though the map will contain as a key some range [C, +∞) for a
+constant C, it should be valid up to some point V > C - and behaviour at timestamps after V may subsequently be changed.
+
+We attempt to read the latest version of the range map, and if our timestamp falls within the validity bound, we
+retrieve the version. If it does not, we submit an identity transformation, which extends the validity bound of the
+current value. We then read the value again, and keep trying until we know what version to use. If we read a version
+we don't support yet, we throw (presumably, this is part of a rolling upgrade and should resolve itself soon).
+
+Installing transactions2 itself is done by a background thread that submits a transform that installs version 2 (or
```

Comment:
continuously? I suppose I think of "installation" as a one time thing, as opposed to something that's continuously running, not sure how else to describe.

Reply:
The thread runs continuously, though there's some logic to not apply the transform if the current version is already 2. I can rephrase this to something like

```
There is a background thread that continuously runs, polling the AtlasDB runtime configuration. If the configuration is configured to a specific version (which at time of writing can be either 1 or 2), it will run a transform setting the transactions schema version going forward to that version. This is optional; if not configured, we will not attempt to run any transforms.
```

Reply:
sounds good.